python/mspasspy/db/database.py

```diff
@@ -574,45 +645,182 @@ def save_data(self, mspass_object, mode="promiscuous", storage_mode='gridfs', di
+                            copied_metadata.erase(k)
+            # Done editing, now we convert copied_metadata to a python dict
+            # using this Metadata method or the long version when in cautious or pedantic mode
+            insertion_dict = dict()
+            if mode == "promiscuous":
+                # A python dictionary can use Metadata as a constructor due to
+                # the way the bindings were defined
+                insertion_dict = dict(copied_metadata)
+            else:
+                # Other modes have to test every key and type of value
+                # before continuing.  pedantic kills data with any problems
+                # Cautious tries to fix the problem first
+                # Note many errors can be posted - one for each problem key-value pair
+                for k in copied_metadata:
+                    if save_schema.is_defined(k):
+                        if not isinstance(copied_metadata[k], save_schema.type(k)):
+                            insertion_dict[k]=copied_metadata[k]
+                        else:
+                            if mode == 'pedantic':
+                                mspass_object.kill()
+                                message ='pedantic mode error:  key='+k
+                                if not save_schema.is_defined(k):
+                                    message += ' not defined in schema.'
+                                else:
+                                    value=copied_metadata[k]
+                                    message += ' type of stored value='+str(type(value))+' does not match schema expectation='+str(save_schema.type(k))
+                                mspass_object.elog.log_error('Database.save_data',
+                                    'message',ErrorSeverity.Invalid)
+                            else:
+                                # Careful if another mode is added here.  else means cautious in this logic
+                                 try:
+                                     # The following convert the actual value in a dict to a required type.
+                                     # This is because the return of type() is the class reference.
+                                     insertion_dict[k] = save_schema.type(k)(copied_metadata[k])
+                                 except Exception as err:
+                                     mspass_object.kill()
+                                     message = 'cautious mdoe error:  key=' + k
+                                     if save_schema.is_required(k):
+                                        message += ' Required key value could not be converted to required type='+str(save_schema.type(k))+' actual type='+str(type(copied_metadata[k]))
+                                     else:
+                                        message += ' Value has type='+str(type(copied_metadata[k])+' which cannot be converted to required type='+str(save_schema.type(k)))
```

Comment:
minor issue here in the message, in the else condition, it should remove the word 'required'.

Reply:
Agreed, done